File: Logic Building Problems/Basic Problems/Closest Number/C++/itterative_appoach.cpp
/*
Iterative Checking - O(m) Time and O(1) Space
The basic idea is to start checking from n - m to n + m one by one and take the closest number.
*/

#include <iostream>
#include <climits>

int closestNumber(int n, int m) {
    int closest = 0, minimumDiffernce = INT_MAX;

    for (int i = n - abs(m); i <= n + abs(m); i++)
        if (!(i % m)) {
            int difference = abs(n - i);

            if (difference < minimumDiffernce || (difference == minimumDiffernce && abs(i) > abs(closest))) {
                closest = i;
                minimumDiffernce = difference;
            }
        }
    
    return closest;
}

int main() {
    int n = -15, m = 6;

    printf("%d", closestNumber(n, m));

    return 0;
}

File: Logic Building Problems/Basic Problems/Closest Number/Python/itterative_appoach.py
"""
Iterative Checking - O(m) Time and O(1) Space
The basic idea is to start checking from n - m to n + m one by one and take the closest number.
"""

def closest_number(n: int, m: int) -> int:

    minimum_differnce = float("inf")
    closest = 0

    for i in range(n - abs(m), n + abs(m) + 1):
        if not i % m:
            difference = abs(n - i)

            if difference < minimum_differnce or (difference == minimum_differnce and abs(i) > abs(closest)):
                closest = i
                minimum_differnce = difference
    
    return closest


if __name__ == "__main__":
    n: int = -15
    m: int = 6

    print(closest_number(n, m))
